File: find-peak-element/find-peak-element.cpp
// class Solution {
// public:
//     int findPeakElement(vector<int>& arr) {
//         int lo=0,hi=arr.size()-1;
//         while(lo<hi-1){
//             int mid=lo+(hi-lo)/2;
//             if(arr[mid]<arr[hi])
//                 lo=mid+1;
//             else if(arr[mid]>arr[hi])
//                 hi=mid;
//         }
//         return hi;
//     }
// };

class Solution {
public:
    int findPeakElement(vector<int>& a) {
        int low=0,n=a.size(),high=n-1,mid;
        if(n==1) return 0;
        while(low<=high)
        {
            mid=low+(high-low)/2;
            if(mid==0 && a[mid]>a[mid+1]) return mid;
            if(mid==n-1 && a[mid-1]<a[mid]) return mid;
            if(mid<n-1 && a[mid]>a[mid+1] && a[mid]>a[mid-1]) return mid;
            if(mid>0 && a[mid]<a[mid-1]) high=mid-1;
            else low=mid+1;
        }
        return -1;
    }
};

File: 162-find-peak-element/find-peak-element.py
class Solution:
    def findPeakElement(self, arr: List[int]) -> int:
        l = 0
        r = len(arr) - 1  # Adjust the right boundary to len(arr) - 1

        # Edge case: if there's only one element or the first element is a peak
        if r+1 == 1 or arr[0] > arr[1]:
            return 0
        
        # Edge case: if the last element is a peak
        if arr[-1] > arr[-2]:
            return r
        
        while l <= r:
            mid = (l + r) // 2

            if arr[mid] < arr[mid + 1]:  # Peak is to the right
                l = mid + 1
            else:  # Peak is to the left or is at mid
                r = mid - 1

        # At the end of the loop, l will be the peak index
        return l
